lib/Dialect/Pipeline/PipelineOps.cpp

```diff
@@ -123,14 +139,44 @@ void ScheduledPipelineOp::build(OpBuilder &odsBuilder, OperationState &odsState,
   entryBlock.addArguments(
       extInputs.getTypes(),
       llvm::SmallVector<Location>(extInputs.size(), odsState.location));
+  // entry stage valid signal.
+  entryBlock.addArgument(odsBuilder.getIntegerType(1), odsState.location);
 }
 
 Block *ScheduledPipelineOp::addStage() {
   OpBuilder builder(getContext());
   Block *stage = builder.createBlock(&getRegion());
+
+  // Add the stage valid signal.
+  stage->addArgument(builder.getIntegerType(1), getLoc());
   return stage;
 }
 
+void ScheduledPipelineOp::getAsmBlockArgumentNames(
+    mlir::Region &region, mlir::OpAsmSetValueNameFn setNameFn) {
+  for (auto [i, block] : llvm::enumerate(getRegion())) {
+    if (&block == getEntryStage()) {
+      size_t nRegularInputs = getInputs().size();
+      size_t nExternalInputs = getExtInputs().size();
+      for (auto [argi, arg] :
+           llvm::enumerate(block.getArguments().slice(0, nRegularInputs)))
+        setNameFn(arg, "s0_arg" + std::to_string(argi));
```

Comment:
I'm pretty sure there's an llvm function call to avoid `std::to_string`. Also, I think that this results in a `std::string` concatenation which returns a new `std::string`. If `setNameFn` supports it, please use `Twine` instead.